pydantic/json_schema.py

```diff
@@ -492,6 +492,12 @@ def handler_func(schema_or_field: CoreSchemaOrField) -> JsonSchemaValue:
             if self.mode == 'serialization' and 'serialization' in schema_or_field:
                 ser_schema = schema_or_field['serialization']  # type: ignore
                 json_schema = self.ser_schema(ser_schema)
+                if (
+                    json_schema is not None
+                    and ser_schema.get('when_used') in ('unless-none', 'json-unless-none')
+                    and schema_or_field['type'] == 'nullable'
+                ):
+                    json_schema = self.get_flattened_anyof([{'type': 'null'}, json_schema])
```

Comment:
I'm not too happy with how this is done. Ideally, `self.nullable_schema` should be used but it will use `schema['schema']` to generate the inner schema:

https://github.com/pydantic/pydantic/blob/a42fb6776ca46e3a3099f868f66ba6635a5e1d5a/pydantic/json_schema.py#L1084-L1094

While we need `schema['serialization']` to be used instead.

Reply:
Should we model the below more closely?

```py
if (
    self.mode == 'serialization' 
    and 'serialization' in schema_or_field
    and (ser_schema := schema_or_field['serialization']
    and (json_schema := self.ser_schema(ser_schema))
    and ser_schema.get('when_used') in ('unless-none', 'json-unless-none')
    and schema_or_field['type'] == 'nullable'
):
    json_schema = self.get_flattened_anyof([{'type': 'null'}, json_schema])
```

Or something like that

Reply:
I understand your concern, but I agree, I think this is a fine patch for now. Maybe add a comment explaining the pattern here - we have so many conditionals, it's a bit hard to understand all together 👍

Reply:
> Should we model the below more closely?

I think I prefer the two conditionals, as it shows the logic more clearly imo:

- if we are in serialization mode and a serialization schema is available:
  - shortcut JSON Schema generation and use `ser_schema`
  - special case `nullable` schema depending on the value of `when_used`